spec/versions_file_spec.rb

```diff
@@ -0,0 +1,101 @@
+
+describe BundlerApi::GemInfo do
+  let(:db)       { $db }
+  let(:builder)  { GemBuilder.new(db) }
+  let(:versions_file) { BundlerApi::VersionsFile.new(db) }
+  let (:yesterday) { DateTime.now - 86400 }
+  let (:tomorrow) { DateTime.now + 86400 }
+  let (:next_week) { DateTime.now + (7*86400) }
+  let (:file_creation_time) { Time.now }
+
+  before do
+    a = builder.create_rubygem("a")
+    b = builder.create_rubygem("b")
+    builder.create_version(a, "a", "0.0.1", "ruby", true, yesterday)
+    builder.create_version(b, "b", "0.0.2", "ruby", true, yesterday)
+    builder.create_version(b, "b", "0.1.1", "java", true, yesterday)
+    builder.create_version(b, "b", "0.1.2", "ruby", true, yesterday)
+  end
+  let (:file_contents) { "0123456789\n---\na 0.0.1\nb 0.0.2,0.1.1-java,0.1.2" }
+
+  describe "#create"  do
+    it "create the versions.list file with the gems on database" do
+      file = Tempfile.new('versions.list')
+      with_versions_file file.path do
+        versions_file.create
+        file.rewind
+        expect(file.read).to match(/\d+\n---\na 0\.0\.1\nb 0\.0\.2,0\.1\.1-java,0\.1\.2/)
+      end
+    end
+  end
+
+  describe "#update" do
+    before do
+      b = builder.rubygem_id("b")
+      c = builder.create_rubygem("c")
+      builder.create_version(b, "b", "0.2.0", "rbx", true, tomorrow)
+      builder.create_version(b, "b", "0.2.0", "ruby", true, tomorrow)
+      builder.create_version(c, "c", "1.0.0", "ruby", true, tomorrow)
+    end
+
+    it "add new versions to the bottom of file" do
+      file = Tempfile.new('versions.list')
+      file.write file_contents
+      file.rewind
+      with_versions_file file.path do
+        versions_file.update
+        file.rewind
+        expect(file.read).to eq(file_contents + "\nb 0.2.0,0.2.0-rbx\nc 1.0.0")
+      end
+    end
+  end
+
+  describe "#with_new_gems" do
+    context "is has nothing new" do
+      before { File.any_instance.stub(:read).and_return("file_contents") }
+      before { File.any_instance.stub(:mtime).and_return(file_creation_time) }
+
+      it "return the same content from versions.list file" do
+        expect(versions_file.with_new_gems).to eq(versions_file.send(:content))
+      end
+    end
+
+    context "when has something new" do
+      before { File.any_instance.stub(:read).and_return(file_contents) }
+      before { File.any_instance.stub(:mtime).and_return(file_creation_time) }
+      before do
+        b = builder.rubygem_id("b")
+        c = builder.create_rubygem("c")
+        builder.create_version(b, "b", "0.2.0", "rbx", true, tomorrow)
+        builder.create_version(b, "b", "0.2.0", "ruby", true, tomorrow)
+        builder.create_version(c, "c", "1.0.0", "ruby", true, tomorrow)
+      end
+
+      it "return the content from versions.list with new gems on bottom" do
+        expect(versions_file.with_new_gems).to eq(file_contents + "\nb 0.2.0,0.2.0-rbx\nc 1.0.0")
+      end
+
+      context "from different dates" do
```

Comment:
The web server runs on many machines at the same time. They cannot share the file, so we cannot update the file itself.

We also want the file to be short when generated in advance--one line for all versions of one name in the generated file. Then, we want to add to the file append-only, which means one line per new version. With append only, the file can be cached on clients for months, and just updated with a few new lines when needed.

On Jun 5, 2015, 8:53 PM +0800, Felipe Tanus<notifications@github.com>, wrote:
>  
> Inspec/versions_file_spec.rb(https://github.com/bundler/bundler-api/pull/78#discussion_r31809377):
>  
> >+ context "when has something new" do>+ before { File.any_instance.stub(:read).and_return(file_contents) }>+ before { File.any_instance.stub(:mtime).and_return(file_creation_time) }>+ before do>+ b = builder.rubygem_id("b")>+ c = builder.create_rubygem("c")>+ builder.create_version(b, "b", "0.2.0", "rbx", true, tomorrow)>+ builder.create_version(b, "b", "0.2.0", "ruby", true, tomorrow)>+ builder.create_version(c, "c", "1.0.0", "ruby", true, tomorrow)>+ end>+>+ it "return the content from versions.list with new gems on bottom" do>+ expect(versions_file.with_new_gems).to eq(file_contents + "\nb 0.2.0,0.2.0-rbx\nc 1.0.0")>+ end>+>+ context "from different dates" do
>  
> Oh, well, then it is even easier :-)
>  
>  
> But then, why don't we
> 1) Append the gem and version on the file every time we would add the gem version in our database
> 2) Serve the file as a static file
>  
>  
> —
> Reply to this email directly orview it on GitHub(https://github.com/bundler/bundler-api/pull/78/files#r31809377).
>  
>  
>